2018-08-06-toykio.md

```diff
@@ -0,0 +1,195 @@
+---
+layout: post
+title:  "Toykio"
+author: "Alexander Polakov"
+author_github: "polachok"
+date:   2018-08-01
+categories: blog
+---
+
+In this blog post I'd like to present [toykio](https://github.com/polachok/toykio), a simple futures executor intended for learning about how executors with an event loop work. Toykio only provides a very minimal feature set: An event loop and TCP streams and listners. However, it turns out that due to the fact that futures are composable, this is enough to build complex clients and servers.
+
+In the following, I'd like to give you a quick overview of toykio's components.
+
+## `AsyncTcpStream`
+
+Tokio defines the `AsyncTcpStream` type, a newtype wrapper around `TcpStream` from the standard library. Just like in std, the `connect` function opens a connection, and sets the socket to non-blocking mode. This means that the `read()` and `write()` methods will return immediately, either with data or an error. If there's not enough data (for reads) or buffer space (for writes) avaialble, a special kind of error - `WouldBlock` - is returned. We'll discuss how to handle it in the next section.
+
+## `AsyncRead` & `AsyncWrite`
+The `AsyncRead` and `AsyncWrite` traits are the foundation of all I/O futures. The `AsyncReadExt` and `AsyncWriteExt`
+extension methods (like `read` and `write_all`) are built on top of them. These traits provide a way to connect futures to an event loop while keeping them independent from any particular event loop implementation.
+
+Let's take a look at the [AsyncRead](https://github.com/polachok/toykio/blob/futures-0.3/src/async_tcp_stream.rs#L42) implementation for `AsyncTcpStream`:
+
+```rust
+impl AsyncRead for AsyncTcpStream {
+    fn poll_read(&mut self, cx: &mut Context, buf: &mut [u8]) -> Poll<Result<usize, Error>> {
+        match self.0.read(buf) {
+            Ok(len) => Poll::Ready(Ok(len)),
+            Err(ref err) if err.kind() == std::io::ErrorKind::WouldBlock => {
+                // Get TcpStream file descriptor
+                let fd = self.0.as_raw_fd();
+                let waker = cx.waker();
+
+                REACTOR.with(|reactor| reactor.add_read_interest(fd, waker.clone()));
+
+                Poll::Pending
+            }
+            Err(err) => panic!("error {:?}", err),
+        }
+    }
+}
+```
+It tries to perform a read on the underlying `TcpStream` and if the read succeeds, the slice will be filled with data. If however the read fails with a `WouldBlock` error, then it registers its interest so that it gets notified when the data becomes available. More about this in the next section.
```

Comment:
s/"registers its interest so that it gets notified"/"registers the current task's waker so that it gets awoken"